modules/caddyhttp/httpredirectlistener.go

```diff
@@ -0,0 +1,120 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package caddyhttp
+
+import (
+	"bufio"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+)
+
+// httpRedirectListener is listener that checks the first few bytes
+// of the request when the server is intended to accept HTTPS requests,
+// to respond to an HTTP request with a redirect.
+type httpRedirectListener struct {
+	net.Listener
+}
+
+// NewHttpRedirectListener wraps inner, handling
+func NewHttpRedirectListener(inner net.Listener) net.Listener {
+	l := &httpRedirectListener{Listener: inner}
+	return l
+}
+
+// Accept waits for and returns the next connection to the listener,
+// wrapping it with a httpRedirectConn.
+func (l *httpRedirectListener) Accept() (net.Conn, error) {
+	c, err := l.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	return &httpRedirectConn{
+		Conn: c,
+		r:    bufio.NewReader(c),
+	}, nil
+}
+
+type httpRedirectConn struct {
+	net.Conn
+	once sync.Once
+	r    *bufio.Reader
+}
+
+// Read tries to peek at the first few bytes of the request, and if we get
+// an error reading the headers, and that error was due to the bytes looking
+// like an HTTP request, then we perform a HTTP->HTTPS redirect on the same
+// port as the original connection.
+func (c *httpRedirectConn) Read(p []byte) (int, error) {
+	var errReturn error
+	c.once.Do(func() {
+		firstBytes, err := c.r.Peek(5)
+		if err != nil {
+			return
+		}
+
+		// If the request doesn't look like HTTP, then it's probably
+		// TLS bytes and we don't need to do anything.
+		if !firstBytesLookLikeHTTP(firstBytes) {
+			return
+		}
+
+		// Parse the HTTP request, so we can get the Host and URL to redirect to.
+		req, err := http.ReadRequest(c.r)
+		if err != nil {
+			return
+		}
+
+		// Build the redirect response, using the same Host and URL,
+		// but replacing the scheme with https.
+		headers := make(http.Header)
+		headers.Add("Location", "https://"+req.Host+req.URL.String())
+		resp := &http.Response{
+			Proto:      "HTTP/1.0",
+			Status:     "308 Permanent Redirect",
+			StatusCode: 308,
+			ProtoMajor: 1,
+			ProtoMinor: 0,
+			Header:     headers,
+		}
+
+		err = resp.Write(c.Conn)
+		if err != nil {
+			errReturn = fmt.Errorf("couldn't write HTTP->HTTPS redirect")
+			return
+		}
+
+		errReturn = fmt.Errorf("redirected HTTP request on HTTPS port")
+		c.Conn.Close()
+	})
+
+	if errReturn != nil {
+		return 0, errReturn
+	}
+
+	return c.r.Read(p)
+}
+
+// firstBytesLookLikeHTTP reports whether a TLS record header
+// looks like it might've been a misdirected plaintext HTTP request.
+func firstBytesLookLikeHTTP(hdr []byte) bool {
+	switch string(hdr[:5]) {
+	case "GET /", "HEAD ", "POST ", "PUT /", "OPTIO":
```

Comment:
I don't love this. It's missing PATCH and a bunch of other verbs that happen over the HTTP port (think WebDAV, etc.). Can we instead check to see if it matches the start of a TLS handshake?

Reply:
I just copied this from the stdlib: 

https://github.com/golang/go/blob/2b39d86344608423138b648b98157470d3809ee7/src/net/http/server.go#L3605-L3613

Reply:
Hmm, ok. If that's the case, we can probably drop the number of bytes to 3 and check only for GET, HEA, POS, and OPT. We might still want to add QUE if QUERY becomes a thing.

Reply:
I'm not sure I see the benefit of that. It would just increase the risk of mismatching the request with other protocols that might also start with those bytes. I think sticking to doing it the same was as stdlib has a greater advantage than we'd get by only comparing against two less bytes.

Reply:
I still think checking for a TLS handshake specifically is a better solution than hard-coding some HTTP methods. I don't know of any other protocols browsers use that start with those byte sequences, do you?

Reply:
Checking for it not being TLS would mean we might respond with HTTP to a request that _wasn't_ HTTP (i.e any other garbage request). It only makes sense to respond with HTTP if we actually received HTTP.

> I don't know of any other protocols browsers use that start with those byte sequences, do you?

Not that I know of, but I also can't predict the future, nothing's stopping anyone from writing a plain-text protocol that ends up looking sorta similar to HTTP :man_shrugging: 

Obviously we're talking hypotheticals at this point, but I think there's precedent for just yoinking logic from stdlib and trusting that it makes sense. We don't need to try to be any more clever here than we're already trying to be :sweat_smile:.

Reply:
The error from stdlib we're trying to preempt with this code is here:

https://github.com/golang/go/blob/2b39d86344608423138b648b98157470d3809ee7/src/net/http/server.go#L1848-L1859

Which explicitly does this exact check, peek the first 5 bytes and see if it's HTTP, and if it is then write a super basic HTTP repsonse `"Client sent an HTTP request to an HTTPS server."`.

We're just _doing the same check_ but slightly earlier so that we can do a _better_ response to this than just saying "nope"

Reply:
Ok, you make some valid points for looking for HTTP specifically. However, I think following the logic that this is mainly to benefit web browsers, we can nuke PUT and probably should add QUERY (which seems likely to become standardized).

Reply:
Can we just wait until Go stdlib makes that change themselves? Again, I find it really unlikely that anything but GET (and maybe, I guess, in edgecases, POST) ever matter here. I don't think we should put `QUERY` in here until it's actually accepted in an RFC. It would just be bitrot otherwise, if it gets rejected.

Reply:
Sure, as long as you help maintain it :wink: (which you're very good about, and I appreciate that).